src/charm.py

```diff
@@ -1,69 +1,159 @@
+        mongo_config["replication"]["replSetName"] = "rs0"
+        with open("/etc/mongod.conf", "w") as mongo_config_file:
+            yaml.dump(mongo_config, mongo_config_file)
+
+        self.unit.status = ops.model.ActiveStatus()
+
+    def _on_start(self, event: ops.charm.StartEvent) -> None:
+        """Enables MongoDB service and initializes replica set.
+
+        Args:
+            event (ops.charm.StartEvent): The triggering start event.
+        """
+        if not self.unit.is_leader():
+            return
+
+        # start mongo service
+        self.unit.status = MaintenanceStatus("enabling MongoDB")
+        if not service_running("mongod.service"):
+            logger.debug("enabling MongoDB")
+            mongod_enabled = service_resume("mongod.service")
+            if not mongod_enabled:
+                logger.error("failed to enable mongod.service")
+                self.unit.status = BlockedStatus("failed to enable mongod.service")
+                return
+
+        try:
+            self._open_port_tcp(self._port)
+        except Exception:
+            self.unit.status = BlockedStatus("failed to open TCP port for MongoDB")
+            return
+
+        if not self.mongo.is_ready():
+            self.unit.status = WaitingStatus("Waiting for MongoDB Service")
+            event.defer()
+            return
+
+        # initialize replica set
+        self.unit.status = MaintenanceStatus("initializing MongoDB replicaset")
+        logger.debug("initalizing replica set")
+        try:
+            logger.debug("initialzing replica set for these IPs %s", self._unit_ips)
+            self.mongo.initialize_replica_set(self._unit_ips)
+            self._peers.data[self.app]["replica_set_hosts"] = json.dumps(self._unit_ips)
+        except Exception as e:
+            logger.error("Error initializing replica sets in _on_start: error={}".format(e))
+            self.unit.status = BlockedStatus("failed to initialize replicasets")
+            return
+
+        self.unit.status = ActiveStatus("MongoDB started")
+
+    def _open_port_tcp(self, port: int) -> None:
+        """Open the given port.
+        
+        Args:
+            port: The port to open.
+        """
+        try:
+            check_call(["open-port", f"{port}/TCP"])
+        except subprocess.CalledProcessError as e:
+            logger.exception(f"failed opening port {port}", exc_info=e)
+            raise Exception(f"failed to open port {port}")
+
+    def _add_mongodb_org_repository(self) -> None:
+        """Adds MongoDB repo to container."""
         repositories = apt.RepositoryMapping()
 
         # Get GPG key
         try:
             key = urlopen("https://www.mongodb.org/static/pgp/server-5.0.asc").read().decode()
         except Exception as e:
-            logger.exception(f"failed to get GPG key", exc_info=e)
+            logger.exception("failed to get GPG key, reason: %s", e)
             self.unit.status = BlockedStatus("Failed to get GPG key")
```

Comment:
```suggestion
            self.unit.status = BlockedStatus("couldn't install MongoDB")
```

Try to keep these messages a little more high level, drop more detail in the logs (as you have already :))